### PR TITLE
Add interactive tests

### PR DESCRIPTION
**Pull request by ellbosch**
This proposes a set of changes to help improve code coverage by over a full percentage point (local run has us at 69.8%).

Changes include:
* Introduction of interactive tests
* Consolidated testing logic between interactive and non-interactive tests
* Changes errors outputted to console to exception raises, which improves testability and thread cleanup
* Commented out an entire unused method (needed for a future update)

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -152,8 +152,8 @@ def unit_test():
     utility.exec_command(('pytest -l --cov mssqlcli --doctest-modules '
                           '--junitxml=junit/test-{}-results.xml --cov-report=xml '
                           '--cov-report=html --cov-append -o junit_suite_name=pytest-{} '
-                          '-s -m "not unstable" {}').format(runid, python_version,
```
Comment (ellbosch) [inline build.py]: Turns out we don't need `-s` flag anymore.

File: mssqlcli/mssql_cli.py [hunk 2]
```diff
@@ -390,15 +389,13 @@ def run(self):
...
-            raise ValueError("'run' must be used in interactive mode! Please set \
-                             interactive_mode to True.")
+            raise ValueError("Invalid arguments: 'run' must be used in interactive mode! Please set "
+                             "interactive_mode to True.")
 
         # exit and return error if user enters interactive mode with -o argument enabled
         if self.output_file:
-            click.secho("Invalid arguments: -o must be used with interactive mode set to false.",
-                        err=True, fg='red')
-            self.shutdown()
```
Comment (ellbosch) [inline mssqlcli/mssql_cli.py]: Raising errors completes shutdown for us—I've validated that threads will not leak.

File: tests/test_noninteractive_mode.py [hunk 3]
```diff
@@ -24,20 +28,14 @@ def tmp_filepath():
         yield fp
         os.remove(fp)
 
-    testdata = [
-        ("-Q \"SELECT 1\"", 'small.txt'),
-        ("-Q \"SELECT 1; SELECT 2;\"", 'multiple.txt'),
-        ("-Q \"SELECT %s\"" % ('x' * 250), 'col_too_wide.txt'),
-        ("-Q \"SELECT REPLICATE(CAST('X,' AS VARCHAR(MAX)), 1024)\"", 'col_wide.txt')
-    ]
```
Comment (ellbosch) [inline tests/test_noninteractive_mode.py]: Moved the above code to util class.

File: tests/test_noninteractive_mode.py [hunk 4]
```diff
@@ -63,35 +62,23 @@ def test_output_file(self, query_str, test_file, tmp_filepath):
...
-    def test_noninteractive_run():
-        """ Test that calling run throws an exception only when interactive_mode is false """
-        mssqlcli = create_mssql_cli(interactive_mode=False)
-        try:
-            mssqlcli.run()
-            assert False
-        except ValueError:
-            assert True
-        finally:
-            mssqlcli.shutdown()
```
Comment (ellbosch) [inline tests/test_noninteractive_mode.py]: Moved above method to util class.

File: tests/test_noninteractive_mode.py [hunk 5]
```diff
@@ -100,13 +87,6 @@ def test_Q_with_i_run(cls):
         output = cls.execute_query_via_subprocess("-Q 'select 1' -i 'this_breaks.txt'")
         assert output == "Invalid arguments: either -Q or -i may be specified."
 
-    @staticmethod
-    def input_output_paths(test_file_suffix):
-        """ Returns tuple of file paths for the input an output of a test. """
-        i = os.path.join(_BASELINE_DIR, 'test_query_inputs', 'input_%s' % test_file_suffix)
-        o = os.path.join(_BASELINE_DIR, 'test_query_baseline', 'baseline_%s' % test_file_suffix)
-        return (i, o)
-
```
Comment (ellbosch) [inline tests/test_noninteractive_mode.py]: Moved method to util.

File: tests/test_noninteractive_mode.py [hunk 6]
```diff
@@ -122,19 +102,9 @@ def execute_query_via_subprocess(cls, query_str, output_file=None):
...
-    @staticmethod
-    def get_file_contents(file_path):
-        """ Get expected result from file. """
-        try:
-            with open(file_path, 'r') as f:
-                # remove string literals (needed in python2) and newlines
-                return f.read().replace('\r', '').strip()
-        except OSError as e:
-            raise e
-
```
Comment (ellbosch) [inline tests/test_noninteractive_mode.py]: Moved method to util.

**Comment by chlafreniere** [comment]
Where does one find code coverage #'s for mssql-cli? Do we have this automated yet?

**Comment by ellbosch** [comment]
> Where does one find code coverage #'s for mssql-cli? Do we have this automated yet?

Yes, all automated in ADO. In addition to 'Code Coverage' tabs which are available in each build run, we have numbers in our [ADO dashboard](https://dev.azure.com/mssqltools/cli/_dashboards/dashboard/32cf741a-1003-49a8-b058-38ffa5378b50).

**Pull request merged**